statsmodels/formula/_manager.py

```diff
@@ -0,0 +1,723 @@
+            return formulaic.Formula(
+                lhs=formulaic.Formula(lhs_formula, _ordering="none"),
+                rhs=formulaic.Formula(intercept + cats + final_conts, _ordering="none"),
+                _ordering="none",
+            )
+        else:
+            return formulaic.Formula(intercept + cats + final_conts, _ordering="none")
+
+    def get_matrices(
+        self,
+        formula,
+        data,
+        eval_env=0,
+        pandas=True,
+        na_action=None,
+        prediction=False,
+    ) -> (
+        np.ndarray
+        | tuple[np.ndarray, np.ndarray]
+        | pd.DataFrame
+        | tuple[pd.DataFrame, pd.DataFrame]
+    ):
+        """
+        Get the design matrix and response vector from the formula and data.
+
+        Parameters
+        ----------
+        formula
+        data
+        eval_env
+        pandas
+        na_action
+
+        Returns
+        -------
+
+        """
+        if isinstance(eval_env, (int, np.integer)):
+            eval_env = int(eval_env) + 1
+        if self._engine == "patsy":
+            return_type = "dataframe" if pandas else "matrix"
+            kwargs = {}
+            if na_action:
+                kwargs["NA_action"] = na_action
+            if isinstance(eval_env, Mapping):
+                _eval_env = patsy.eval.EvalEnvironment(
+                    [{key: val} for key, val in eval_env.items()]
+                )
+            else:
+                _eval_env = eval_env
+                if isinstance(eval_env, patsy.eval.EvalEnvironment):
+                    warnings.warn(EVAL_ENV_WARNING, FutureWarning)
+            if (
+                isinstance(
+                    formula, (patsy.design_info.DesignInfo, patsy.desc.ModelDesc)
+                )
+                or "~" not in formula
+                or formula.strip().startswith("~")
+            ):
+                output = patsy.dmatrix(
+                    formula, data, eval_env=eval_env, return_type=return_type, **kwargs
+                )
+            else:  # "~" in formula:
+                output = patsy.dmatrices(
+                    formula, data, eval_env=eval_env, return_type=return_type, **kwargs
+                )
+            if isinstance(output, tuple):
+                self._spec = output[1].design_info
+            else:
+                self._spec = output.design_info
+            if isinstance(na_action, NAAction):
+                self._missing_mask = getattr(na_action, "missing_mask", None)
+            return output
+
+        else:  # self._engine == "formulaic":
+            import statsmodels.formula
+
+            kwargs = {}
+            if pandas:
+                kwargs["output"] = "pandas"
+            else:
+                kwargs["output"] = "numpy"
+
+            if na_action:
+                kwargs["na_action"] = na_action
+
+            if isinstance(data, (dict, Mapping)):
+                # Work around for no dict support in formulaic
```

Comment:
I might as well solve this from Formulaic's side. I intentionally didn't accept ambiguous data types, but pragmatically, we should likely do our best to just function with these common data-types.

Reply:
Thanks for adding this.